skimage/util/dask_compatibility.py

```diff
@@ -0,0 +1,493 @@
+and Dask (``da``) namespaces. Ideally this module would be imported and used
+as follows::
+
+>>> import dask_compatibility as dc
+>>> dc.rint(arr)  # Identical behavior if arr is from Dask or NumPy
+"""
+
+__all__ = ['asanyarray',
+           'asarray',
+           'clip',
+           'dtype_conversion_inplace',
+           'eager_min',
+           'eager_max',
+           'empty',
+           'est_best_chunk',
+           'floor',
+           'maximum',
+           'rint',
+           ]
+
+
+def _graceful_compute(variable):
+    """
+    Attempts to call .compute() on every item in variable.
+    """
+    # Shortcut on NumPy arrays first
+    if isinstance(variable, np.ndarray):
+        # Return the NumPy array
+        return variable
+
+    # Dask arrays are not iterable, but NumPy arrays are.
+    elif not isinstance(variable, collections.Iterable):
+        # Could be a Dask array, try to compute it
+        try:
+            return variable.compute()
+        except AttributeError:
+            # Not a Dask array, return the variable.
+            return variable
+
+    # Might be multiple values in a container
+    else:
+        # TODO: graceful computing of items in dicts
+        try:
+            # Attempt to gracefully compute each item
+            return tuple(_graceful_compute(x) for x in variable)
+        except TypeError:
+            # If neither of these work it isn't a Dask array; return as-is
+            return variable
+
+
+def _dask_input(*args, **kwargs):
+    """
+    Determines if any inputs contain Dask arrays.
+
+    Parameters
+    ----------
+    *args, **kwargs
+        All arguments from the enclosing function should be passed.
+
+    Returns
+    -------
+    dask_present : bool
+        True if at least one input was a Dask array.
+    """
+    return (any(isinstance(x, da.core.Array) for x in args) or
+            any(isinstance(x, da.core.Array)
+                for _, x in six.iteritems(kwargs)))
+
+
+def dask_decorator(compute=True):
+    """
+    Decorator to control conversion of outputs from Dask to NumPy arrays
+
+    Parameters
+    ----------
+    compute : bool
+        Decorator argument allows the default behavior (True, do the
+        conversion) to be overridden so Dask arrays can be passed along
+        to another function.
+
+    Returns
+    -------
+    decorator : function
+        Decorated function, ready for use.
+    """
+    def real_decorator(function):
+        def wrapper(*args, **kwargs):
+            if _dask_input(*args, **kwargs) or not compute:
+                return function(*args, **kwargs)
+            else:
+                return _graceful_compute(function(*args, **kwargs))
+        return wrapper
+    return real_decorator
+
+
+def _total_elements(shape):
+    """
+    Efficiently determine the number of elements in an array, given its shape.
+    """
+    return functools.reduce(operator.mul, shape, 1)
```

Comment:
The idea was to be able to find a fairly-reasonable chunk that fits in memory _before_ making the array, so `.size` wouldn't yet be available.